subsys/jaxrs/src/main/java/org/commonjava/indy/bind/jaxrs/MDCManager.java

```diff
@@ -0,0 +1,102 @@
+package org.commonjava.indy.bind.jaxrs;
+
+import org.apache.http.Header;
+import org.apache.http.HttpRequest;
+import org.commonjava.indy.conf.IndyConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.MDC;
+
+import javax.annotation.PostConstruct;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.servlet.http.HttpServletRequest;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+import static org.apache.commons.lang.StringUtils.isNotBlank;
+
+import static org.commonjava.indy.bind.jaxrs.RequestContextConstants.CLIENT_ADDR;
+import static org.commonjava.indy.bind.jaxrs.RequestContextConstants.COMPONENT_ID;
+import static org.commonjava.indy.bind.jaxrs.RequestContextConstants.EXTERNAL_ID;
+import static org.commonjava.indy.bind.jaxrs.RequestContextConstants.INTERNAL_ID;
+import static org.commonjava.indy.bind.jaxrs.RequestContextConstants.PREFERRED_ID;
+
+@ApplicationScoped
+public class MDCManager
+{
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    @Inject
+    private IndyConfiguration config;
+
+    public MDCManager() {}
+
+    public void clear()
+    {
+        MDC.clear();
+    }
+
+    private List<String> mdcHeadersList = new ArrayList();
+
+    @PostConstruct
+    public void setUp()
+    {
+        mdcHeadersList.add( COMPONENT_ID ); // default
+
+        String mdcHeaders = config.getMdcHeaders();
+        if ( isNotBlank( mdcHeaders ) )
+        {
+            String[] headers = mdcHeaders.split( "," );
+            for ( String s : headers )
+            {
+                if ( isNotBlank( s ) && !mdcHeadersList.contains( s ) )
+                {
+                    mdcHeadersList.add( s.trim() );
+                }
+            }
+        }
+    }
+
+    public void putExternalID( String externalID )
+    {
+        if ( externalID == null )
+        {
+            logger.debug( "No externalId" );
+            return;
+        }
+        String internalID = UUID.randomUUID().toString();
+        String preferredID = externalID != null ? externalID : internalID;
```

Comment:
It looks like internalID will never be used here, because of the return statement above...am I missing something?

Reply:
This internalID is used in the next line - calc the preferredId.